### PR TITLE
Gen-Plugin - Fixes for 3rd Party Integration, and Entrypoints

### LINKED ISSUE #418: Gen-Plugin for 3rd Party Plugins generates an incorrect entrypoint
Gen-Plugin for 3rd Party Plugins generates an incorrect script entrypoint.

When a 3rd party plugin is generated, installed, and the `aac -h` command is run the following error appears:
```
Traceback (most recent call last):
  File "/workspace/AaC/python/venv/bin/aac", line 33, in <module>
    sys.exit(load_entry_point('aac', 'console_scripts', 'aac')())
  File "/workspace/AaC/python/src/aac/cli/execute.py", line 16, in run_cli
    arg_parser, aac_plugin_commands = _setup_arg_parser()
  File "/workspace/AaC/python/src/aac/cli/execute.py", line 47, in _setup_arg_parser
    aac_and_plugin_commands = get_plugin_commands()
  File "/workspace/AaC/python/src/aac/plugins/plugin_manager.py", line 89, in get_plugin_commands
    return list(flatten([plugin.get_commands() for plugin in get_plugins() if plugin.get_commands()]))
  File "/workspace/AaC/python/src/aac/plugins/plugin_manager.py", line 79, in get_plugins
    return get_plugin_manager().hook.get_plugin()
  File "/workspace/AaC/python/src/aac/plugins/plugin_manager.py", line 22, in get_plugin_manager
    plugin_manager.load_setuptools_entrypoints(PLUGIN_PROJECT_NAME)
  File "/workspace/AaC/python/venv/lib/python3.9/site-packages/pluggy/_manager.py", line 287, in load_setuptools_entrypoints
    plugin = ep.load()
  File "/usr/lib/python3.9/importlib/metadata.py", line 77, in load
    module = import_module(match.group('module'))
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 984, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'test_plugin.test_plugin'
```

The expected entrypoint in `setup.py.jinja2` should be:
```python
"aac": ["{{plugin.name}}={{plugin.implementation_name}}"],
```

The pipeline stage to extend can be found here: https://github.com/jondavid-black/AaC/blob/fe9909a43e0bf5ed33426947a5e6198b65fecedd/.github/workflows/python-tests.yml#L56-L61

## AC:
- [x] Gen-Plugin generates 3rd party plugin entrypoints correctly
- [x] Gen-Plugin moves 3rd party plugin yaml files into the generated plugin files
- [ ] The 3rd Party Plugin integration test attempts to run `aac -h` after the 3rd party plugin is installed
  - [ ] The stage must fail before the plugin is fixed
  - [ ] The state passes after the plugin is fixed 
  - [ ] There is a new 3rd party plugin yaml model file that is distinct from any 1st party plugins
  - [ ] The new plugin file command is present in the help output after installation

### PR DESCRIPTION
**Pull request by JustAChucklin**
PR closes #418 

Addresses the following issues:

1. Invalid entrypoint for the 3rd party plugins
2. Architecture file for the plugin not being in the plugin folder within the root folder of the plugin
  2.1. E.g. Architecture file now goes to the folder where the `{plugin-name}.impl.py` file is
3. Added some additional jobs to the `Third-Party` integration test to test for the new third party plugin to validate it was installed properly.

### CODE REVIEW

File: python/src/aac/plugins/gen_plugin/gen_plugin_impl.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import yaml
 
 from os import path
+from os import rename
```
Comment (cvchaparro) [inline python/src/aac/plugins/gen_plugin/gen_plugin_impl.py:11]:
```suggestion
from os import path, rename
```

File: python/src/aac/plugins/gen_plugin/gen_plugin_impl.py [hunk 2]
```diff
@@ -350,3 +353,14 @@ def _get_repository_root_directory_from_path(system_path: str) -> str:
         )
 
     return system_path[:target_index]
+
+
+def _move_architecture_file_to_plugin_root(architecture_file_path: str, plugin_name: str, output_directory: str) -> None:
+
```
Comment (cvchaparro) [inline python/src/aac/plugins/gen_plugin/gen_plugin_impl.py:359]:
```suggestion
```

File: python/src/aac/plugins/gen_plugin/gen_plugin_impl.py [hunk 3]
```diff
@@ -171,7 +172,7 @@ def _prepare_and_generate_plugin_files(
     template_output_directories = _get_template_output_directories(
         plugin_type, architecture_file_path, plugin_implementation_name
     )
-
+    _move_architecture_file_to_plugin_root(architecture_file_path, plugin_implementation_name, output_directory)
```
Comment (cvchaparro) [inline python/src/aac/plugins/gen_plugin/gen_plugin_impl.py:175]: This should probably go inside the `_generate_plugin_files_to_directory` function after `write_generated_templates_to_file`.

**Review by cvchaparro — CHANGES_REQUESTED** [review]

File: python/src/aac/plugins/gen_plugin/gen_plugin_impl.py (hunk 3 above)
Comment (cvchaparro) [reply python/src/aac/plugins/gen_plugin/gen_plugin_impl.py:175]: FYSA: `write_generated_templates_to_file` eventually calls `write_file` which will [create directories if they don't exist](https://github.com/jondavid-black/AaC/blob/974578f1102e9f0edca55bb2caef685db93d2059/python/src/aac/io/writer.py#L24-L25) and, until that's called the plugin directory may not exist.

**Comment by JustAChucklin** [comment]
I think that I addressed your concerns @cvchaparro

File: .github/workflows/python-tests.yml [hunk 4]
```diff
@@ -53,13 +54,15 @@ jobs:
...
         pip install .[all]
-
+    
     - name: Test Gen-Plugin Third-Party Plugin
       run: |
         echo y | aac gen-plugin model/plugin/plugin.yaml
         cd model/plugin/
-        pip install .[all]
-
+        pip install .:w
```
Comment (cvchaparro) [inline .github/workflows/python-tests.yml:62]: What's the purpose of `:w`?

File: .github/workflows/python-tests.yml [hunk 5]
```diff
@@ -53,13 +54,15 @@ jobs:
         cp model/plugin/plugin.yaml src/aac/plugins/first_party/first_party.yaml
         echo y | aac gen-plugin src/aac/plugins/first_party/first_party.yaml
         pip install .[all]
-
+
```
Comment (cvchaparro) [inline .github/workflows/python-tests.yml:57]: Watch out for whitespace here.

```suggestion

```

File: .github/workflows/python-tests.yml [hunk 6]
```diff
@@ -53,13 +54,15 @@ jobs:
...
+    
     - name: Test Gen-Plugin Third-Party Plugin
       run: |
         echo y | aac gen-plugin model/plugin/plugin.yaml
         cd model/plugin/
-        pip install .[all]
-
+        pip install .:w
+        
+        aac -h | grep "test-plugin-command" && echo 'Succefully installed Third-Party plugin!'
```
Comment (cvchaparro) [inline .github/workflows/python-tests.yml:64]: Not needed.

```suggestion
        aac -h | grep "test-plugin-command"
```

File: .github/workflows/python-tests.yml [hunk 7]
```diff
@@ -86,66 +89,67 @@ jobs:
         run: echo "::set-output name=wheel_name::$(basename $(ls dist/*.whl))"
 
   os_test_matrix:
-      name: Python ${{ matrix.os }} Tests
```
Comment (cvchaparro) [inline .github/workflows/python-tests.yml:89]: These spacing changes may cause pipeline issues (YAML cares about spacing). Can you please revert everything in this file that you didn't specifically mean to change so that it's clear what you did mean to change?

File: python/model/test_plugin/test_plugin.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+model:
```
Comment (cvchaparro) [inline python/model/test_plugin/test_plugin.yaml:1]: Why did you add this file instead of using `model/plugin/plugin.yaml`. I don't think there's anything 1st-/3rd-party-specific in that file so the same plugin file should work for both versions of the command.

File: python/src/aac/io/parser/_parse_source.py [hunk 9]
```diff
@@ -37,7 +37,7 @@ def parse(source: str, source_uri: Optional[str] = None) -> list[Definition]:
 
     if linesep not in source:
         sanitized_source = sanitize_filesystem_path(source)
-        if path.lexists(sanitized_source):
+        if path.exists(sanitized_source):
```
Comment (cvchaparro) [inline python/src/aac/io/parser/_parse_source.py:40]: Why did you change this?

File: python/src/aac/plugins/gen_plugin/templates/third_party/setup.py.jinja2 [hunk 10]
```diff
@@ -24,12 +24,13 @@ test_dependencies = [
 setup(
     version="0.0.1",
     name="{{plugin.name}}",
+    package_data={"": ["*.aac", "*.jinja2", "*.yaml"]},
```
Comment (cvchaparro) [inline python/src/aac/plugins/gen_plugin/templates/third_party/setup.py.jinja2:27]: This is untested but, according to [this](https://docs.python.org/3/distutils/setupscript.html#installing-package-data), I think this may be what you want:

```suggestion
    packages=["{{plugin.name}}"],
    package_dir={"{{plugin.name}}": "src/{{plugin.name}}"},
    package_data={"{{plugin.name}}": ["*.aac", "*.jinja2", "*.yaml"]},
```

If that doesn't work, you may need to change the references to `{{plugin.name}}` with `{{plugin.implementation_name}}` but it should be pretty easy to test each of them to see what works.

**Review by cvchaparro — CHANGES_REQUESTED** [review]

File: python/src/aac/plugins/gen_plugin/templates/third_party/setup.py.jinja2 (hunk 10 above)
Comment (JustAChucklin) [reply python/src/aac/plugins/gen_plugin/templates/third_party/setup.py.jinja2:27]: I have this in my local but it didnt get pushed up for some reason

Comment (JustAChucklin) [reply python/src/aac/plugins/gen_plugin/templates/third_party/setup.py.jinja2:27]: sorry I thought they were in local but was missing some of my previous implementation. Should be pushing up here soon.

File: .github/workflows/python-tests.yml (hunk 6 above)
Comment (JustAChucklin) [reply .github/workflows/python-tests.yml:64]: Removed the additional command

File: .github/workflows/python-tests.yml (hunk 5 above)
Comment (JustAChucklin) [reply .github/workflows/python-tests.yml:57]: Removed

File: .github/workflows/python-tests.yml (hunk 4 above)
Comment (JustAChucklin) [reply .github/workflows/python-tests.yml:62]: Editor quirp, should be removed now.

File: python/model/test_plugin/test_plugin.yaml (hunk 8 above)
Comment (JustAChucklin) [reply python/model/test_plugin/test_plugin.yaml:1]: Gotcha
Will remove

File: python/src/aac/io/parser/_parse_source.py (hunk 9 above)
Comment (JustAChucklin) [reply python/src/aac/io/parser/_parse_source.py:40]: This was a change that was made during troubleshooting, when there was a suspected issue with the parser. 

@Coffee2Bits , do you think this could be changed back? since we seemed to have found the issue root to be related to the `io.py helper`

File: python/src/aac/plugins/gen_plugin/templates/third_party/setup.py.jinja2 (hunk 10 above)
Comment (JustAChucklin) [reply python/src/aac/plugins/gen_plugin/templates/third_party/setup.py.jinja2:29]: alright changes should pushed through...I think during my local testing I found that the `plugin.implementation_name` seemed to play better than the `plugin.name` due to the file structure was named from the former.

**Comment by JustAChucklin** [comment]
Ok I think this fully ready for review now. @cvchaparro @Coffee2Bits

File: python/src/aac/plugins/gen_plugin/templates/third_party/setup.py.jinja2 (hunk 10 above)
Comment (JustAChucklin) [reply python/src/aac/plugins/gen_plugin/templates/third_party/setup.py.jinja2:28]: The `packages=find_packages(where=".", exlude="tests")` is what got it going.

File: python/src/aac/plugins/gen_plugin/gen_plugin_impl.py (hunk 3 above)
Comment (JustAChucklin) [reply python/src/aac/plugins/gen_plugin/gen_plugin_impl.py]: I think this should be good now. Let me know if otherwise.

File: python/src/aac/io/parser/_parse_source.py (hunk 9 above)
Comment (Coffee2Bits) [reply python/src/aac/io/parser/_parse_source.py:40]: Yes, I would change it back to `lexists`. The change during our debug session was just a quick check to make sure that there wasn't a filesystem issue that might.

**Review by Coffee2Bits — CHANGES_REQUESTED** [review]
Good work, almost there! Just left a couple suggestions for edits.

File: .github/workflows/python-tests.yml [hunk 11]
```diff
@@ -124,13 +122,11 @@ jobs:
           run: |
             spec_file=$(find $pythonLocation -name spec.yaml | head -n1)
             python -m aac validate $spec_file
-
         - name: Validate Core Spec
           if: matrix.os == 'windows-latest'
           run: |
             $spec_file=$(gci -r -fi spec.yaml -Name "$Env:pythonLocation")
             python -m aac validate "$Env:pythonLocation\$spec_file"
-
```
[the remark this answers is not in the document]
Comment (JustAChucklin) [reply .github/workflows/python-tests.yml:133]: Should be pushing those changes through now

File: python/src/aac/io/parser/_parse_source.py (hunk 9 above)
Comment (JustAChucklin) [reply python/src/aac/io/parser/_parse_source.py:40]: Alright I will add it back

Comment (JustAChucklin) [reply python/src/aac/io/parser/_parse_source.py]: Alright the changes have been reverted.

File: .github/workflows/python-tests.yml [hunk 12]
```diff
@@ -46,20 +46,18 @@ jobs:
...
     - name: Test Gen-Plugin Third-Party Plugin
       run: |
         echo y | aac gen-plugin model/plugin/plugin.yaml
         cd model/plugin/
-        pip install .[all]
-
+        pip install .
+        aac -h | grep "test-plugin-command"
     - name: Upload logs from failed runs
```
[the remark this answers is not in the document]
Comment (JustAChucklin) [reply .github/workflows/python-tests.yml:64]: I think this has been changed back in the latest round of commits let me know if they are needing to be changed further

**Review by Coffee2Bits — APPROVED** [review]
Looks good! Great job!

I suspect you'll need to get Cameron to sign off on his review as well in order to merge.

**Comment by JustAChucklin** [comment]
> Looks good! Great job!
> 
> I suspect you'll need to get Cameron to sign off on his review as well in order to merge.

Thanks for checking it out again. I will check in with Cameron today, regarding the finalizing of his review.

File: python/tests/plugins/test_gen_plugin.py [hunk 13]
```diff
@@ -59,20 +59,23 @@ def test_generate_first_party_plugin(self, is_user_desired_output_dir):
...
+            assert_plugin_success(result)
 
-                generated_plugin_files = os.listdir(temp_directory)
-                self.assertEqual(len(generated_plugin_files), 6)
-                self.assertIn(os.path.basename(plugin_yaml.name), generated_plugin_files)
-                self.assertIn(f"{TEST_PLUGIN_NAME}", generated_plugin_files)
+            generated_plugin_files = []
+            for root, dirs, files in os.walk(temp_directory, topdown=False):
+                for file in files:
+                    generated_plugin_files.append(os.path.join(root, file))
```
Comment (cvchaparro) [inline python/tests/plugins/test_gen_plugin.py:75]: A few things:

1. Comprehensions are a more concise way to build this list.
2. No real need for `topdown=False` here.
3. For variables that aren't used, it's idiomatic to use `_` - so I changed `dirs` to `_`.
4. Looking at the [tests in pipeline run](https://github.com/jondavid-black/AaC/actions/runs/3200458262/jobs/5227412915#step:5:39) I see the following "ignored" exception which seems to indicate that something with these tests are still not working correctly (this also happens locally):
```
Exception ignored in: <function _TemporaryFileCloser.__del__ at 0x7fca0bdf0310>
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.0/x64/lib/python3.9/tempfile.py", line 441, in __del__
    self.close()
  File "/opt/hostedtoolcache/Python/3.9.0/x64/lib/python3.9/tempfile.py", line 437, in close
    unlink(self.name)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmp0m5mvk_9/tmp854d6e2p.yaml'
```

```suggestion
            generated_plugin_files = [os.path.join(root, file) for root, _, files in os.walk(temp_directory) for file in files]
```

**Review by cvchaparro — CHANGES_REQUESTED** [review]
I have a few comments - reach out if you have any questions. Also, please be sure to set up your editor to remove whitespace at the end of lines.

File: python/tests/plugins/test_gen_plugin.py (hunk 13 above)
Comment (JustAChucklin) [reply python/tests/plugins/test_gen_plugin.py:75]: I applied the following above snippet locally and the tests are still failing. 

```......Exception ignored in: <function _TemporaryFileCloser.__del__ at 0x104078f70>
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.9/3.9.14/Frameworks/Python.framework/Versions/3.9/lib/python3.9/tempfile.py", line 445, in __del__
    self.close()
  File "/opt/homebrew/Cellar/python@3.9/3.9.14/Frameworks/Python.framework/Versions/3.9/lib/python3.9/tempfile.py", line 441, in close
    unlink(self.name)
FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/lr/d4cwjct95f31nlq6vc6pxg7m0000gq/T/tmpkok8m09j/tmp807yvdbf.yaml'
.Exception ignored in: <function _TemporaryFileCloser.__del__ at 0x104078f70>
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.9/3.9.14/Frameworks/Python.framework/Versions/3.9/lib/python3.9/tempfile.py", line 445, in __del__
    self.close()
  File "/opt/homebrew/Cellar/python@3.9/3.9.14/Frameworks/Python.framework/Versions/3.9/lib/python3.9/tempfile.py", line 441, in close
    unlink(self.name)
FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/lr/d4cwjct95f31nlq6vc6pxg7m0000gq/T/tmp35kim0l0/tmpftn7f26q'
.```

Comment (cvchaparro) [reply python/tests/plugins/test_gen_plugin.py]: To reiterate my comment in stand up, this change was not intended to fix the `FileNotFoundError` but rather it was intended to clean up the creation of the `generated_plugin_files` list.

**Review by cvchaparro — APPROVED** [review]

**Comment by JustAChucklin** [comment]
Thanks all!
